### ISSUE TITLE
Type errors on WebSocket Factory

### ISSUE THREAD

**Issue by EvanCarroll** [ohjames/rxjs-websockets#48]
```
index.ts:39:2 - error TS2345: Argument of type '(x: string, y: string | string[] | undefined) => WebSocket' is not assignable to parameter of type 'WebSocketFactory | undefined'.
  Type '(x: string, y: string | string[] | undefined) => WebSocket' is not assignable to type 'WebSocketFactory'.
    Type 'WebSocket' is not assignable to type 'IWebSocket'.
      Types of property 'onopen' are incompatible.
        Type '(event: { target: WebSocket; }) => void' is not assignable to type '((event: Event) => any) | undefined'.
          Type '(event: { target: WebSocket; }) => void' is not assignable to type '(event: Event) => any'.
            Types of parameters 'event' and 'event' are incompatible.
              Type 'Event' is not assignable to type '{ target: WebSocket; }'.
                Types of property 'target' are incompatible.
                  Type 'EventTarget | null' is not assignable to type 'WebSocket'.
                    Type 'null' is not assignable to type 'WebSocket'.

39  (x,y) => new WebSocket(x,y)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

The reason is that if you use `ws`, it's typed with [`@types/ws`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ws/index.d.ts) like this,

    onopen: (event: { target: WebSocket }) => void;
    onerror: (event: {error: any, message: string, type: string, target: WebSocket }) => void;
    onclose: (event: { wasClean: boolean; code: number; reason: string; target: WebSocket }) => void;
    onmessage: (event: { data: WebSocket.Data; type: string; target: WebSocket }) => void;

While currently the factory is typed with an interface like this,

    onopen?: (event: Event) => any
    onclose?: (event: CloseEvent) => any
    onmessage?: (event: MessageEvent) => any
    onerror?: (event: ErrorEvent) => any

Moreover, [ErrorEvent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/preloadjs/index.d.ts) and the like are typed like this,

````
export class ErrorEvent
    {
    constructor(title?: string, message?: string, data?: Object);

    // properties
    data: Object;
    message: string;
    title: string;
    }
````

I think we should mirror the definitions of `ws` rather than define our own. I assume `ws` is mirroring the definition of the browsers native types (when provided).

**Comment by ohjames** [comment]
The types being used are the native types, ws is the one doing it's own thing. ws should be updated rather than this library.

**Comment by ohjames** [comment]
I think the types for this library also need to improved though... maybe ws is onto something.

**Comment by EvanCarroll** [comment]
We either have to be *stronger* typed accepting **only** things that provide `ws`. Or, massively weaker typed to the common denominator between `ws` and w3c. At that point we're pretty much just dropping TypeScript support explicitly with `any`.

We can do that too if you want me to strip the types. As it is though, I don't think the factory thing works in a strongly typed system unless any contemporary node.js modules are producing the w3c types, certainly not `ws` which is the big one (afaik).

Not sure how other client/server code bases do it. Maybe request.js has some wisdom on it?

**Comment by ohjames** [comment]
I'm fixing this now, sorry it took so long.

**Comment by ohjames** [comment]
@EvanCarroll Linked the fixing commit to this issue, annoying I had to use `any`, not sure why typescript doesn't apply function bivariance when considering type compatibility of an object's properties to an interface. Released `v6.0.4` with the fix.
